### ISSUE TITLE
[yardoc] Can't skip parsing README.md

### ISSUE THREAD

**Issue by jvalencia80** [lsegal/yard#837]
I've tried several ways of skipping the **README.md** file without success.

My **.yardopts** is right now:

    --exclude README.md
    lib/**/*.rb

Every single time, a file called **file.README.html** is created, and what is worse, **index.html** will render it by default. I also remove the **doc** directory every time, so the file is not a remanent from the last invocation.

The parser is working, according to this:

````
$ yardoc --debug
[debug]: Parsing ["lib/**/*.rb"] with `ruby` parser
[debug]: Parsing lib/aio.rb
[debug]: Parsing lib/aio/board.rb
[debug]: Parsing lib/aio/sketch.rb
[debug]: Parsing lib/aio/errors.rb
[debug]: Parsing lib/aio/helpers.rb
[debug]: Parsing lib/aio/constants.rb
[debug]: Serializing to .yardoc/objects/root.dat
...
````

Is this a bug? How can I skip it?

**Comment by lsegal** [comment]
The `--exclude` parameter applies to source files, not "extra files" (what YARD calls non-source doc files like READMEs).

Why are you trying to ignore the README from your generated documentation?

**Comment by jvalencia80** [comment]
Because I wanted the API documentation to be just a reference to the methods. I wanted the README to appear on Github only.

**Comment by lsegal** [comment]
Presumably if you wrote a README file, you've done so because it is important documentation for your project. READMEs are important, and there are no plans to support the use case of ignoring explicitly written documentation from documentation-- marking this as WontFix.

You're free to reassemble the generated HTML after YARD has created it (a simple `mv _index.html index.html` would do) or even write your own template customizations if you want to.

**Comment by martinb3** [comment]
:+1: because README.md has special meaning to Github, and I'd also like to show one thing to the Github audience but document my code for another audience. If you won't fix, you should probably rename `--exclude` to something that more accurately reflects its purpose, such as `--exclude_source`.

**Comment by lsegal** [comment]
@martinb3 you can set the `--readme` parameter to point to your custom README file. This specific issue is asking to skip the README file altogether, which is not supported. We can update the documentation to be more detailed (patches accepted) but we can't rename a highly relied on option name without a breaking change.

**Comment by ginty** [comment]
Well, I have the exact same wish as @martinb3 and this seems like a perfectly reasonable scenario to support.
Speaking for myself, I want the Github audience to be presented with an introduction to my project, I want my API audience to be presented with an API index.

**Comment by ginty** [comment]
Why can't a --[no-]readme option be added, that's not a breaking change right?

**Comment by ginty** [comment]
For anyone else frustrated by this, the comment by @lsegal above about moving _index.html is a good solution. _index.html is the index you would get if you didn't have a README, so you can just copy it over.
However that page now contains a files section at the top with a link to a README that no longer exists.
If that annoys you and you never have any other files, you can remove it like this:

~~~ruby
system("yard doc --output-dir #{Origen.root}/web/output/api --title '#{title}'")
# Yard doesn't have an option to ignore github-style READMEs, so force it here to
# always present the API index on the API homepage for consistency
index = "#{Origen.root}/web/output/api/index.html"
_index = "#{Origen.root}/web/output/api/_index.html"
FileUtils.rm_f(index) if File.exist?(index)
require 'nokogiri'
doc = Nokogiri::HTML(File.read(_index))
doc.xpath('//h2[contains(text(), "File Listing")]').remove
doc.css("#files").remove
File.open(_index, 'w') {|f| f.write(doc.to_html) }
FileUtils.cp(_index, index)
~~~

So a big hack, but that basically gives the same end result as --no-readme would.

**Comment by najamelan** [comment]
Why is this closed? 

This doesn't only concern the main README.md. The fact that --exclude does not take into account the extra files is strange, counter intuitive and problematic. If I have a project with git submodules and I --exclude them for the documentation generation, I will still get a README file with no title in my yardoc output for every submodule. That's a no go, and sure everything is open source, we can rewrite templates and parsers and everything else, but why?

**Comment by lsegal** [comment]
@najamelan the explanation for closure was provided above, specifically,

> Presumably if you wrote a README file, you've done so because it is important documentation for your project. READMEs are important, and there are no plans to support the use case of ignoring explicitly written documentation from documentation-- marking this as WontFix.

One of YARD's goals as a project is to make documentation better. The reason that READMEs are generated in your project root isn't a GitHub specific things, it's because READMEs are an important part of documentation in any project. Removing a README from your documentation does no help to improve a project's documentation. If your README files do not have pertinent information, you may not want to have them around-- and if they do have pertinent information, then you want them in your documentation, because a README by definition is documentation.

Hope that explains it.

**Comment by lsegal** [comment]
@najamelan I should add, it sounds like your issue might be slightly different. YARD only generates a single README for any project. If you're running into an issue where your other README files are being included as extra files, it sounds like those files are coming from sub-directories, and you're using an overly inclusive wildcard for extra files. YARD does not pull these READMEs in by default (it only pulls from the root), so this would be something opted into by your `.yardopts`. 

One workaround would be to list your extra files more specifically (don't use overly global wildcards). That said, in your case, you may benefit from an `--exclude-extra-file` type flag (since `--exclude` is only for code files), and if you want to submit a PR for `--exclude-extra-file`, I'd be happy to look at it. That said, neither `--exclude` or `--exclude-extra-file` would ever remove the core README, only subsequent READMEs from subdirs.

**Comment by najamelan** [comment]
I have indeed in the mean time reduced my wildcard. It's counter intuitive and more work though to have to list all of them separately instead of being able to use a wildcard and expect that folders I exclude would just not be visited by yard at all. Unfortunately my develop time is limited, so I won't be able to contribute to yard at this moment. 

As a side note, I consider it a good programming principle to leave decisions up to the end user and not to limit their options. You can never foresee all the possible use cases of a tool, so the less doors you close the more people can use it. 

> One of YARD's goals as a project is to make documentation better. 

What if a user decides to put a concise readme on github to point to the online documentation or with instructions on running yard, where they want another readme to be the main page of their documentation. Who would be legitimate in deciding that nobody should ever do this, and thus make it impossible? 

People look for tools, not for moral lessons, so if you fail to appreciate that, they turn elsewhere. You won't force the hand of a user who knows what they want. Trying to do so is patronizing in my opinion, and that's why people get aggravated over it. There's a number of otherwise reasonably good tools that I won't use for that exact reason, websites blocking tor, rspec or minitest, ... Do you have a saw that auto-destructs after a certain number of uses because the makers decided that sawing with a blunt saw is bad for your back? I don't think you'd want one, well neither do I...

**Comment by lsegal** [comment]
> As a side note, I consider it a good programming principle to leave decisions up to the end user and not to limit their options.

Conversely, opinionated software is fairly common in the Ruby ecosystem. Fortunately, YARD actually makes fairly few opinionated choices, so I feel confident in standing behind the very few decisions we do make. That said, it's still possible to do the thing needed in this issue with minor template tweaks, so to say that we limit options is actually fairly inaccurate to begin with. It is more accurate to say that we make certain things hard to do, because they should be.

As for your example use case,

> What if a user decides to put a concise readme on github to point to the online documentation or with instructions on running yard, where they want another readme to be the main page of their documentation. 

This is already possible with `--readme`. The specific issue here is not wanting any README document at all, which simply does not fall under YARD's goals.

> Do you have a saw that auto-destructs after a certain number of uses because the makers decided that sawing with a blunt saw is bad for your back? I don't think you'd want one, well neither do I...

YARD is not a saw that auto-destructs, YARD is a table saw that [does this](https://youtu.be/ZCgtwRq9C8c?t=170). That *is* a table saw I would want, which is why I build similar safety mechanisms into the software I write. If you prefer a saw that will slice off your hand, you are correct that there certainly are other options.

**Comment by thomthom** [comment]
> This is already possible with --readme. 
Is it possible with that directive to make YARD to the default index page? Because that is what I'm trying to do as well; a README for GitHub on instructions for building the project etc. Though still have the documentation index be the class overview.

**Comment by lsegal** [comment]
@thomthom no, `--readme` must point to a file (currently). You should really consider a *user* README.

It's fine if you want to consider your "GitHub README" as developer-focused information on building and contributing to a body of code, but that doesn't preclude you from having a user-focused landing page that introduces users to your API (ideally that's what your GitHub README should be, though I understand that different projects might have different requirements). Think of YARD's "readme" as a landing page for users reading your documentation. In this context, users still need to know what the project is, what they should expect from the API documentation, and how to get started using the project.

The index / overview is not an appropriate landing page for documentation. The index is [just that](https://www.reference.com/education/index-book-d1cb900c7666130a), a reference of classes for quick lookup by users who already know what they want to look for. It's a visual search optimization, and nothing more. An index, just like in traditional publishing, should sit at the back of your content, not the front.

tl;dr YARD's index is not a landing page. Documentation needs a landing page in the same way that code on GitHu needs a landing page. Make a landing page for your documentation.

**Comment by thomthom** [comment]
We do actually have a landing page for our API - but that is part of a larger developer site. Than that then points to the API docs. So that's why we want the YARD generated docs to dive straight into the index page. Part of this is that we have APIs in other languages than Ruby, generated with other tools. So our API docs site isn't built solely by YARD - it's just a small part of it.

**Comment by thomthom** [comment]
FYI to anyone interested, it is possible to make GitHub use a README other than the one in the root of the repository; Create a `.github/README.md` file and GitHub will prioritise that instead.

**Comment by tayjaybabee** [comment]
> I have indeed in the mean time reduced my wildcard. It's counter intuitive and more work though to have to list all of them separately instead of being able to use a wildcard and expect that folders I exclude would just not be visited by yard at all. Unfortunately my develop time is limited, so I won't be able to contribute to yard at this moment.
> 
> As a side note, I consider it a good programming principle to leave decisions up to the end user and not to limit their options. You can never foresee all the possible use cases of a tool, so the less doors you close the more people can use it.
> 
> > One of YARD's goals as a project is to make documentation better.
> 
> What if a user decides ...
>
> People look for tools, not for moral lessons, so if you fail to appreciate that, they turn elsewhere. You won't force the hand of a user who knows what they want. Trying to do so is patronizing in my opinion, and that's why people get aggravated over it. 

I agree with this point and find that a good portion of the simple basic functionality of YARD is entirely far too opinionated. Like the fact that I can't figure out how to make it document code because it's not in a sub-directory of the 'lib' directory so I have to waste my time searching your documentation.

**Comment by lsegal** [comment]
> because it's not in a sub-directory of the 'lib' directory

YARD is simply following a very standardized Ruby convention here, not inventing anything new. It's also worth noting that this same convention will prevent you from using other tools like RDoc, RubyGems, etc, which all default to using `lib` in the same way, so I'm not sure this is a good example of a YARD-specific opinion.

YARD, however, is opinionated, as are most Ruby libraries. If opinions don't match standard conventions, there is room to improve, but I'm not sure that's the case here. 

YARD holds the opinion that a prose based landing page is important for documenting a project. This is most commonly known as the README. This itself shouldn't really be a controversial opinion-- GitHub holds the same opinion, we just conflict on the expected filename. This conflict is due to the difference in audience context for users cloning a repo vs. reading API documentation, frankly an age old conflict that is not unique to Ruby.

Fortunately as mentioned above, you can both change the filename in YARD (`--readme`) or specify an override README in GitHub (`.github/README`) which implies that this is contextual conflict that GitHub is aware of these edge cases. Given the very simple workarounds here, I'm not sure the solution should be a behavioral change. That said, a better highlighting of this in docs could be helpful for some users hitting this edge case.

If the opinion truly is that there should be documentation with no landing page, that is not an opinion that YARD holds, and yes, YARD is opinionated enough to _not help you_ in this case. That said, and as listed above, there are still some very simple workarounds if you really want to work in this way. In that sense, YARD is opinionated but YARD's opinions should not stop you from doing whatever you want with the tool with minimal effort (changing a README file or removing the index page altogether both take only 1 line of code). 

_Discoverability_ is usually the blocker, and that's because these use cases are rarely asked about and so documenting them is not fully prioritized. If you believe discoverability of these edge cases could use work, the best way to fix it would be to add that visibility through a documentation PR, or simply write extra addendum content and open an issue to have it added to YARD's (sparse) set of guides at https://yardoc.org/guides -- I would certainly accept extra docs that help with topics such as "improving documentation on GitHub", "customizing look and feel of docs", and so on.

**Comment by tayjaybabee** [comment]
I have to give you major props for thoroughly answering and addressing this. Makes me want to see what I can do to at least - as you stated - improve education on what opinions YARD (and yes, a good portion of Ruby work) has that might interfere in cases such as mine and OPs

**Comment by DannyBen** [comment]
Although this is an old and closed issue, I feel compelled to comment (apologies for pinging subscribers who no longer care about this issue).

From software author's perspective, I can understand the need to force *some* opinions on your users. Software with unopinionated developers ends up a messy unmaintainable bloatware.

From user perspective, I feel the yard developers have already provided a way to easily work around it, with the `--readme` flag. All it means is that you must have *some* sort of an index.

My `.yardopts` file:

```
--readme support/yardoc-readme.md
```

My yardoc-readme.md file:

```markdown
# Gem Name

This is the documentation for the Gem Name's source code. For the user guide please
refer to ...
```

Crisis averted.

As a side note, earlier in this discussion someone asked for the reason for avoiding the original README - and one answer is that GitHub README's support more features than yardoc, like inline HTML, so the yardoc readme is broken without removing it or replacing it.
